### ISSUE TITLE
Update one-to-many through PATCH on resource not working

### ISSUE THREAD

**Issue by maurei** [json-api-dotnet/JsonApiDotNetCore#492]
A [core feature of the spec](https://jsonapi.org/format/#crud-updating-resource-relationships) is currently not supported:
```json
PATCH /articles/1 HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{
  "data": {
    "type": "authors",
    "id": "1",
    "relationships": {
      "books": {
        "data": [
          { "type": "books", "id": "2" },
          { "type": "books", "id": "3" }
        ]
      }
    }
  }
}
```

**Cause of the problem**
Because of incomplete information on `_jsonApiContext` in the case of `one-to-many`, which is set in the `JsonApiDeserializer`, the `DefaultEntityRepository` does not attempt to update such relationships. Details:

*  In the case of `one-to-one`, the `JsonApiDeserializer` adds meta info about this to `_jsonApiContext.RelationshipsToUpdate` [right here in the SetHasOneRelationship method](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/8f685a6a23515acf8f440c70d20444a0cec1c502/src/JsonApiDotNetCore/Serialization/JsonApiDeSerializer.cs#L214).
* This property however is not updated in the case of `one-to-many`: [in the SetHasManyRelationship method](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/8f685a6a23515acf8f440c70d20444a0cec1c502/src/JsonApiDotNetCore/Serialization/JsonApiDeSerializer.cs#L299), only `_jsonApiContext.HasManyRelationshipPointers` are updated, which is not enough.
*  [this is where the repository](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/8f685a6a23515acf8f440c70d20444a0cec1c502/src/JsonApiDotNetCore/Data/DefaultEntityRepository.cs#L299) would be updating `one-to-many` relationships if they were included in `_jsonApiContext.RelationshipsToUpdate`.
* Also in the repo: `AttachRelationships(oldEntity)` is fired **before** `relationship.Key.SetValue(oldEntity, relationship.Value);` is called, so any updates to the relations are currently not tracked.

A complication: the request above should perform a **complete replace**. This means that before updating the `Author.BookId` to its new value, the foreign keys of all previous `Book`s for that author should be set to null (and an error should be thrown it is non-nullable).

**Solution**
* Add a `hasManyAttr` value to the `RelationshipsToUpdate` dictionary in the `JsonApiDeserializer` [right here](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/8f685a6a23515acf8f440c70d20444a0cec1c502/src/JsonApiDotNetCore/Serialization/JsonApiDeSerializer.cs#L299).
* in the case of a complete replace, include the related entities first [see this topic; same for EF Core](https://stackoverflow.com/questions/30116199/entity-framework-proper-way-to-replace-collection-in-one-to-many)
* do `AttachRelationships(oldEntity)` before actually setting the new relationship collection.

**Note:** the same issue exists for `many-to-many`, but it requires a different fix. Will address this in a separate issue.
## Environment

- JsonApiDotNetCore Version:  latest

**Comment by division-by-zero** [comment]
HasOne relationships are also not updated. This request doesn't update the relationship.

```
PATCH /books/1 HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{
  "data": {
    "type": "books",
    "id": "1",
    "relationships": {
      "author": {
        "data": { "type": "people", "id": "2" }
      }
    }
  }
}
```
This request works though:

```
PATCH /books/1/relationships/author HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{
  "data": { "type": "people", "id": "2" }
}
```

**Comment by maurei** [comment]
There was an additional issue related to entity tracking that caused an error when performing a complete replacement with a new collection that contained an element that was already in a previous collection. This has been addressed in #496.
